### PR TITLE
feat(tracing/internal): keep track of trace sampling decision history

### PR DESCRIPTION
**Pull request by brettlangdon**
We want to keep track of the history of sampling decisions made throughout the lifetime of a trace. This is beneficial to know exactly why a trace was kept or dropped. This is most useful for when we have partial traces, e.g. Service A -> Service B, and somehow the spans from Service A failed to make it to the backend.

We are introducing a new tag to store this information on the trace: `_dd.p.upstream_service`.

This tag must:
- Be parsed from requests from upstream services
- Be propagated to downstream services
- Added as a tag on each chunk root span
- Contain only a single entry for the local trace

The format follows:

```
<base64 encoded span service name>|<sampling priority>|<sampling mechanism>|<sample rate or empty string>[;<other entries>]
```

```
upstream services  =  ( group, { ";", group } ) | "";

group  =  service name, "|",
          sampling priority, "|",
          sampling mechanism, "|",
          sampling rate, { "|", future field };

service name  =  ? unpadded base64 encoded UTF-8 bytes ?;

(* no plus signs, no exponential notation, etc. *)
sampling priority  =  ? decimal integer ? ;

sampling mechanism  =  ? positive decimal integer ?;

sampling rate  =  "" | normalized float;

normalized float  =  ? decimal between 0.0 and 1.0 inclusive with at most four significant digits (rounded up) and with the leading "0." e.g. "0.9877" ?;

(* somewhat arbitrarily based on the other grammar *)
future field  =  ( ? ASCII characters 32-126 ? - delimiter );

delimiter  =  "|" | ","
```

## Challenges

One of the requirements is that we only add one entry per-context. If someone makes a call to manually keep (or keep then drop then keep) we should only keep track of the most recent sampling decision in `_dd.p.upstream_services`.

As of right now we store `_dd.p.upstream_services` in `Context._meta` which makes it very easy (no extra steps needed) to add the tag to the chunk root and propagate between services (#3045), but it makes it difficult to ensure we only have a single entry since we need to ensure that `Context._meta` always has the most up to date version of `_dd.p.upstream_services`.

The way I work around this right now is, the `Context` object will update `_dd.p.upstream_services` with the new entry, and then keep track of the most recent entry. When we go to update the tag again if we know that the most recent entry is ours, then we'll cut it off of the end and append the new one on. The logic for this is a little hairy, and means a decent chunk of overhead around making a sampling decision to partition / append strings.

### CODE REVIEW

File: ddtrace/context.py [hunk 1]
```diff
@@ -74,6 +88,63 @@ def _update_tags(self, span):
...
+                self._meta[UPSTREAM_SERVICES_KEY] = ";".join([existing, entry])
+
+        # Existing upstream value, and none set by us yet
+        elif existing:
+            self._meta[UPSTREAM_SERVICES_KEY] = ";".join([existing, entry])
+        else:
+            self._meta[UPSTREAM_SERVICES_KEY] = entry
+
+        # Keep track of the last entry that we created/added
+        self._last_upstream_service_entry = entry
```
Comment (brettlangdon) [inline ddtrace/context.py:146]: Another approach here could be to keep track of the inherited/extracted `_dd.p.upstream_services` from context propagation separate from "current entry", that way this entire thing just ends up being ("ish"):

```python
self._upstream_service_entry = "|".join([ base_64, priority, mechanism, rate ])
self._meta[UPSTREAM_SERVICES_KEY] = ";".join([
    entry
    for entry in (self._parent_upstream_service_entry, self._upstream_service_entry)
    if entry
])
```

File: ddtrace/sampler.py [hunk 2]
```diff
@@ -313,11 +319,15 @@ def sample(self, span):
...
         else:
             # No rules matches so use agent based sampling
-            if self._agent_sampler.sample(span):
-                self._set_priority(span, AUTO_KEEP)
+            _sampler = self._agent_sampler._get_sampler(span)
+            # TODO: We need to determine the proper SamplingMechanism here
+            # TODO: Do we keep this extracted logic or should we move this logic into RateByServiceSampler.sample ?
+            span.set_metric(SAMPLING_AGENT_DECISION, _sampler.sample_rate)
+            if _sampler.sample(span):
+                self._set_priority(span, AUTO_KEEP, SamplingMechanism.AGENT, _sampler.sample_rate)
```
Comment (brettlangdon) [inline ddtrace/sampler.py:327]: I don't love this.

I think the proper call here is to move what `self._set_priority` is trying to do into a base class for all samplers and then just delegate 100% of this logic to `RateByServiceSampler` / `RateSampler`

It seems like either that or duplicate the `RateByServiceSampler` logic here, which doesn't feel right.

File: ddtrace/sampler.py [hunk 3]
```diff
@@ -327,24 +337,22 @@ def sample(self, span):
         # Sample with the matching sampling rule
         span.set_metric(SAMPLING_RULE_DECISION, matching_rule.sample_rate)
         if not matching_rule.sample(span):
-            self._set_priority(span, USER_REJECT)
+            self._set_priority(span, USER_REJECT, SamplingMechanism.USER_RULE, matching_rule.sample_rate)
             return False
-        else:
-            # Do not return here, we need to apply rate limit
-            self._set_priority(span, USER_KEEP)
```
Comment (brettlangdon) [inline ddtrace/sampler.py:334]: extra call, not needed

File: ddtrace/context.py [hunk 4]
```diff
@@ -74,6 +88,63 @@ def _update_tags(self, span):
...
+
+        # If the sampling decision hasn't changed, do not add an entry
+        # For example, if someone does `span.set_tag(MANUAL_KEEP_KEY)` multiple times in one trace
+        if self._last_upstream_service_entry == entry:
+            return
+
+        # Fetch any existing value (extracted from propagation headers, or we have already set ourselves)
+        existing = self._meta.get(UPSTREAM_SERVICES_KEY)
+
+        # We have an existing header value and a previously set upstream service entry by us
```
Comment (majorgreys) [inline ddtrace/context.py:128]: You have it above but we should say here that any previous entry set by the same instrumented service would be replaced.

**Review by majorgreys — COMMENTED** [review]
some initial comments

File: ddtrace/constants.py [hunk 5]
```diff
@@ -20,6 +23,9 @@
 MANUAL_DROP_KEY = "manual.drop"
 MANUAL_KEEP_KEY = "manual.keep"
 
+# Horizontally propagated tags
```
[the remark this answers is not in the document]
Comment (brettlangdon) [reply ddtrace/constants.py:26]:
```suggestion
# Tags propagated between services
```

File: ddtrace/constants.py [hunk 6]
```diff
@@ -36,3 +42,15 @@
...
 # Use this to explicitly inform the backend that a trace should be kept and stored.
 USER_KEEP = 2
+
+
+class SamplingMechanism(enum.Enum):
+    """The mechanism responsible for making a sampling decision."""
+
+    DEFAULT = 0  # RateByServiceSampler used before receiving rates from the agent
+    AGENT = 1  # RateByServiceSampler used rates from the agent
+    REMOTE = 2  # Not currently used
```
[the remark this answers is not in the document]
Comment (brettlangdon) [reply ddtrace/constants.py:52]:
```suggestion
    REMOTE = 2  # Reserved for future use
```

File: ddtrace/constants.py [hunk 7]
```diff
@@ -36,3 +42,15 @@
...
+    USER_RULE = 3  # Manual rule / rate provided in tracer configuration
+    MANUAL = 4  # Manually set via `set_tag(MANUAL_DROP/KEEP_KEY)`
+    APP_SEC = 5  # Not currently used
+    USER_REMOTE = 6  # Not currently used
```
Comment (brettlangdon) [inline ddtrace/constants.py:56]:
```suggestion
    _ = 5  # Reserved for future use
    _ = 6  # Reserved for future use
```

File: ddtrace/constants.py [hunk 8]
```diff
@@ -36,3 +42,15 @@
 AUTO_KEEP = 1
 # Use this to explicitly inform the backend that a trace should be kept and stored.
 USER_KEEP = 2
+
+
+class SamplingMechanism(enum.Enum):
```
Comment (brettlangdon) [inline ddtrace/constants.py:47]: consider moving this to `ddtrace/internal/constants.py`

File: tests/tracer/test_tracer.py [hunk 9]
```diff
@@ -1365,13 +1365,15 @@ def test_manual_keep(tracer, test_spans):
         s.set_tag(MANUAL_KEEP_KEY)
     spans = test_spans.pop()
     assert spans[0].metrics[SAMPLING_PRIORITY_KEY] is USER_KEEP
+    assert spans[0].context.upstream_services == "dW5uYW1lZC1weXRob24tc2VydmljZQ|2|4|"
```
Comment (brettlangdon) [inline tests/tracer/test_tracer.py:1368]: I couldn't tell if this is the right place for this test/assertion, or if I should add specific manual keep/drop test cases to `test_upstream_services.py`. Open to opinions here.

**Review by brettlangdon — COMMENTED** [review]
~TODO: I probably need to update a _ton_ of snapshot files.~

Done.

File: tests/snapshots/test_multi_trace.json [hunk 10]
```diff
@@ -7,6 +7,7 @@
     "span_id": 1,
```
Comment (brettlangdon) [inline tests/snapshots/test_multi_trace.json:7]: Reviewers: Make sure to scroll to the bottom, there are test changes after this large chunk of snapshot file updates.

**Comment by brettlangdon** [comment]
One thing of note is we are not propagating upstream service entries across forks. Should we?

https://github.com/DataDog/dd-trace-py/blob/92f2a0eeb3bd37fe2375afa0cfdf77b07fd4e7ff/ddtrace/tracer.py#L489-L508

For this we would probably do:

```python
                new_ctx = Context(
                    sampling_priority=child_of.context.sampling_priority,
                    span_id=child_of.span_id,
                    trace_id=child_of.trace_id,
                    upstream_services=child_of.context.upstream_services,
                )
```

**Review by Kyle-Verhoog — COMMENTED** [review]
Just posting my in-flight review

File: ddtrace/internal/utils/upstream_services.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+_ServiceName = Union[Text, bytes]
+
+
+# DEV: The number of services, especially sampling services should be low, usually only 1
+@cached(maxsize=24)
+def _format_service_name(service):
+    # type: (_ServiceName) -> str
+    if not service:
+        service = "unnamed-python-service"
```
[the remark this answers is not in the document]
Comment (brettlangdon) [reply ddtrace/internal/utils/upstream_services.py:20]: Yes, this is part of the spec, it _must_ be this value. I wasn't sure if we should try to standardize or not at this time.

File: ddtrace/internal/_tagset.pyx [hunk 12]
```diff
@@ -152,7 +152,7 @@ cdef bint _value_is_valid(str value):
         return 0
 
     for c in value:
-        if not is_valid_key_char(ord(c)):
+        if not is_valid_value_char(ord(c)):
```
Comment (mabdinur) [inline ddtrace/internal/_tagset.pyx:155]: why do we need to change this method?

File: ddtrace/internal/utils/upstream_services.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mabdinur) [reply ddtrace/internal/utils/upstream_services.py:20]: We define this default service name (unnamed-python-service) in the telemetry writer. Would it make sense to turn this into a constant?

File: ddtrace/internal/_tagset.pyx (hunk 12 above)
Comment (brettlangdon) [reply ddtrace/internal/_tagset.pyx:155]: it is a small bug, catching at the same time since they are related.

Comment (brettlangdon) [reply ddtrace/internal/_tagset.pyx:155]: I can move this to it's own PR if people want. I'm not sure if I'd backport it so felt maybe it was ok?

**Comment by mabdinur** [comment]
Failing test due to ddtrace 1.x: https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/14789/workflows/970fcab8-c890-4177-be24-d569d6108284/jobs/1029787.  tracer arg was removed from Span.__init__()

**Review by mabdinur — DISMISSED** [review]

**Review by mabdinur — DISMISSED** [review]

File: tests/snapshots/tests.contrib.rq.test_rq.test_enqueue_distributed_tracing_enabled_None_worker_service_None.json [hunk 13]
```diff
@@ -0,0 +1,63 @@
+[[
```
[the remark this answers is not in the document]
Comment (brettlangdon) [reply tests/snapshots/tests.contrib.rq.test_rq.test_enqueue_distributed_tracing_enabled_None_worker_service_None.json:1]: this may have been from the conflict resolution, I cannot remember if it was needed in the original PR. I was waiting on tests to re-evaluate

File: ddtrace/sampler.py [hunk 14]
```diff
@@ -137,22 +138,30 @@ def set_sample_rate(
         # type: (...) -> None
         self._by_service_samplers[self._key(service, env)] = RateSampler(sample_rate)
 
-    def _set_priority(self, span, priority):
-        # type: (Span, int) -> None
-        span.context.sampling_priority = priority
+    def _set_priority(self, span, priority, mechanism, sample_rate):
```
[the remark this answers is not in the document]
Comment (brettlangdon) [reply ddtrace/sampler.py:141]: I think that would increase the complexity of the implementation? More code to paths to manage, and since we include the samplers as part of the public API it would be hard to remove/merge them in the future?

File: tests/contrib/rq/test_rq.py [hunk 15]
```diff
@@ -127,9 +127,8 @@ def test_worker_class_job(queue):
     worker.work(burst=True)
 
 
-@pytest.mark.parametrize(
-    "distributed_tracing_enabled,worker_service_name", [(None, "custom-worker-service"), (False, None)]
-)
+@pytest.mark.parametrize("distributed_tracing_enabled", [False, None])
+@pytest.mark.parametrize("worker_service_name", [None, "custom-worker-service"])
```
Comment (brettlangdon) [inline tests/contrib/rq/test_rq.py:131]: @Kyle-Verhoog this is why the new snapshots are there, this small change increases the matrix of tests. probably should be in a different PR, let me know if you want that.

Comment (Kyle-Verhoog) [reply tests/contrib/rq/test_rq.py:131]: yeah can we do that real quick?

File: ddtrace/sampler.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Kyle-Verhoog) [reply ddtrace/sampler.py:141]: Yeah, it does increase the surface area and allow more codepaths but at the same time it is subtly changing the existing implementation. 

Another benefit to introducing a new sampler instead of mutating is that we can do much easier performance testing.

On the flip side there is so much implementation that is shared so I'm ok with just mutating the existing sampler.

File: ddtrace/context.py [hunk 16]
```diff
@@ -29,6 +32,8 @@ class Context(object):
         "_lock",
         "_meta",
         "_metrics",
+        "_upstream_services",
```
Comment (Kyle-Verhoog) [inline ddtrace/context.py:35]: idk if we want to update the `repr` for these new attributes or not

Comment (brettlangdon) [reply ddtrace/context.py:35]: it will show up in meta

Comment (brettlangdon) [reply ddtrace/context.py:35]: remember, these two properties are to maintain the state of "my inherited upstream service entr(y|ies)" and "my upstream service entry", but the current active/correct/propagatable (this should be a word) concatenation of these lives in `_meta`

File: ddtrace/span.py [hunk 17]
```diff
@@ -283,9 +284,19 @@ def set_tag(self, key, value=None):
 
         elif key == MANUAL_KEEP_KEY:
             self.context.sampling_priority = USER_KEEP
+            self.context._update_upstream_services(
```
Comment (Kyle-Verhoog) [inline ddtrace/span.py:287]: Should this code just be added to the getter/setter for sampling_priority? What if someone sets context.sampling_priority manually instead of using set_tag?

File: releasenotes/notes/add-upstream-services-c4fee21ab10bc2fa.yaml [hunk 18]
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    Add support to automatically track and propagate the history of sampling decisions in a distributed trace.
```
Comment (Kyle-Verhoog) [inline releasenotes/notes/add-upstream-services-c4fee21ab10bc2fa.yaml:4]: Not sure how actionable or insightful this is for users. What should they expect to change?

**Review by Kyle-Verhoog — COMMENTED** [review]
Just two more questions but otherwise I think it looks alright to me

File: ddtrace/span.py (hunk 17 above)
Comment (brettlangdon) [reply ddtrace/span.py:287]: Unfortunately we cannot since we need to have the Span (it's service name) when updating this entry.
